Fix crash when using unconfigured default:key

octacian opened the pull request:
Addresses #1640

Since meta secret field is blank on any keys (default:key) obtained using /give, the game will crash if you attempt to use one of these. This follows along the principle that the game should never have any reason to crash.

sofar commented:
<octacian> Early night all! :D
* octacian has quit (Quit: Leaving)

LIES

sofar commented:
I just literally wrote the same patch.

:+1:

sofar commented:
well no, not entirely the same

you are still blindly expecting the `parse_json` to succeed. My version runs `parse_json` and looks at the result of that.

sofar commented:
Note that if the json is bad somehow, you end up with it returning `nil`

`minetest.parse_json(key_oldmeta).secret`

equals

`(nil).secret` 

and we still crash.

sofar commented:
Submitted my version. I slightly prefer it since it eliminates another potential game crash.

octacian commented:
Took a look at your version. I don't see how it eliminates another crash though. It's highly unlikely (in fact, impossible) for the key metadata to be something which `parse_json` wouldn't like. Plus, your method causes an error from `parse_json` to be printed to the chat complaining about how it failed to parse the data. My method fails entirely gracefully (no weird error messages).

If `parse_json` could possibly cause a crash outside of the metadata being blank (if this is the case, please let me know), the best method is to add `and minetest.parse_json(key_oldmeta)` to the if statement in my version. Again, just opinions.

octacian commented:
Whoops, I meant `or not minetest.parse_json(key_oldmeta)` - edited above comment.

octacian commented:
Updated adding `or not minetest.parse_json(key_oldmeta)` to `default.can_interact_with_node` and removed changes in `doors/init.lua` as they conflicted with the recently merged #1580 which changes doors to use the already modified `default.can_interact_with_node` (meaning that I only need to change that one function).

@sofar ^

paramat commented:
07e11427cf86de4554834fe1ebbb72531611774b
